burn-tensor/src/tensor/backend/base.rs

```diff
@@ -3,9 +3,51 @@ use alloc::string::String;
 use crate::ops::*;
 use crate::tensor::Element;
 
+/// This trait defines all types and functions needed for a backend to be used with burn.
+///
+/// ## Design
+///
+/// This trait aims to be as unopinionated as possible and allows implementations to define
+/// their own types and patterns. Therefore, there are few pre-defined abstractions baked
+/// into this trait.
+///
+/// Backends must define their own tensor types for each data type: `float`, `int`, and `bool`.
+/// Since we minimize assumptions, we chose to separate these types, as they are used in
+/// different contexts. However, some backends may have a generic tensor type that is used
+/// for all data types.
+///
+/// ### Eager Mode
+///
+/// Because burn supports dynamic graphs, the backend trait is designed around kernel
+/// implementations that can be called without any mutable context or graph. This may not be
+/// ideal for backends that want to configure their computational graphs and execute them
+/// multiple times.
+///
+/// To implement this kind of backend, channels could be used to communicate with a backend
+/// server thread to build the computation graphs and re-execute the ones that are repeated,
+/// with some form of cache. Once that pattern has matured, a graph mode backend trait could
+/// be extracted from it, allowing other backends of the same kind to be quickly integrated
+/// with burn. This pattern could also be used to create an operation fusion trait, which
+/// allows backends to define what kind of graph structures can be fused into one operation.
+///
+/// ### Multi-Threaded
+///
+/// Backend tensor types are all `Clone` + `Sync` + `Send`, which allows them to be safely
+/// shared between threads. It is recommended to wrap tensors with [Arc](alloc::sync::Arc),
+/// which avoids copying the tensor's buffer. Note that it is still possible to mutate and
+/// reuse tensors' buffer without locking; see the next section on the Mutable API.
+///
+/// ### Mutable API
+///
+/// There is no mutable or inplace operation API to implement, but that does not mean that
+/// backends cannot support them. Using [try_unwrap](alloc::sync::Arc::try_unwrap) and
+/// [get_mut](alloc::sync::Arc::get_mut) allows backends to have access to an owned or mutable
+/// reference to their tensor buffer data structure if the tensor is not shared. In that case,
+/// backends can dispatch to their owned inplace operations for better performance.
```

Comment:
Excellent explanation. I would love more insight like this. Perhaps, Burn deserves it's own Rust style book/documentation.

Reply:
I'll add more and more documentation like this. Maybe a Rust style book at some point will become necessary.